src/chatd.cpp

```diff
@@ -2379,10 +2413,16 @@ void Chat::msgIncomingAfterDecrypt(bool isNew, bool isLocal, Message& msg, Idx i
 
         verifyMsgOrder(msg, idx);
         CALL_DB(addMsgToHistory, msg, idx);
-        if ((msg.userid != mClient.mUserId) &&
-           ((mLastReceivedIdx == CHATD_IDX_INVALID) || (idx > mLastReceivedIdx)))
+        if ((msg.userid != mClient.mUserId)
+                && ((mLastIdxReceivedFromServer == CHATD_IDX_INVALID) || (idx > mLastIdxReceivedFromServer))
+                && ! mClient.chats(mChatId).isGroup())
         {
-            sendCommand(Command(OP_RECEIVED) + mChatId + msgid);
+            mLastIdxReceivedFromServer = idx;
+            mLastIdReceivedFromServer = msgid;
+            if (!isFetchingFromServer())
```

Comment:
What if the incoming message is decrypted while it is fetching more history from server at the same time?